### PR TITLE
Hotp mail

### PR DESCRIPTION
**Pull request by alneberg**
- New endpoint for api to request one-time authentication code (refferred to as HOTP) to be sent via email - Using basic authentication
 - Require valid HOTP to be supplied to the encrypted_token endpoint, otherwise authentication error
 - Remove signed token endpoint
 - New endpoint for web to request HOTP to be sent via email - requires valid username and password
 - Reuquire valid HOTP to be entered in the web login form
 - Moved `dds_web/api/errors.py` to `dds_web/errors.py` to avoid circular imports when using the errors in side `dds_web/database/models.py`

A new HOTP is issued by:
1. Incrementing the hotp counter 
2. Saving the current time

HOTP validation:
1. Check no more than 1 hour has passed since the HOTP was issued
2. verify HOTP code with current counter
3. If verified, increment counter

This should solve #719 and #773

### CODE REVIEW

**Review by inaod568 — COMMENTED** [review]
Could we change slightly to something like this? 
1. Show login form 
2. User fills in form 
3. Submit triggers an email and displays a form which has the username and password as hidden fields, only showing the message (this has been sent to your email) and the field where the code should filled in
4. Submit triggers complete authentication and login. This still means that the authentication occurrs twice but the user doesn't have to fill it in twice. 

Another alternative is to have an extra column called "2fa_authenticated" and that the user gets logged in once the username and password is verified but the `2fa_authenticated` has to be set to True for the user to be able to access any other endpoints.

File: dds_web/utils.py
Comment (talavis) [inline dds_web/utils.py:326]: I can't find `MAIL_SENDER` in `congif.py`. Either add it or set `MAIL_DEFAULT_SENDER` instead, not setting the sender here.

File: dds_web/api/__init__.py
Comment (worukan) [inline dds_web/api/__init__.py:32]: I would have it like 
`api.add_resource(user.SecondFactor, "/user/second_factor", endpoint="second_factor")`
and this endpoint would be used by a user to enter mfa value, not request mfa email.

File: dds_web/api/user.py
Comment (worukan) [inline dds_web/api/user.py:529]: I would keep the encrypted token as it is and only add 
`"message": "Please take this token to /user/second_factor to authenticate with MFA!",`

File: dds_web/api/user.py
Comment (worukan) [inline dds_web/api/user.py:537]: I would replace this class with the following

`class SecondFactor(flask_restful.Resource):
    """Take in and verify an authentication One-Time code entered by an authenticated user with basic credentials"""

    @auth.login_required
    def get(self):
        args = flask.request.json
        if args is None:
            args = {}

        token_schemas.TokenSchema().load(args)

        token_claims = dds_web.security.auth.decrypt_and_verify_token_signature(flask.request.headers["Authorization"].split()[1])

        return flask.jsonify(
            {
                "token": update_token_with_mfa(token_claims)
            }
        )`

File: requirements.txt
Comment (worukan) [inline requirements.txt:38]: where is this used?

**Comment by worukan** [comment]
The intersection for the token and MFA code to work together would be in auth.py
```
...
@auth.verify_token
def verify_token(token):
...
    if dds_web.utils.current_time() <= datetime.datetime.fromtimestamp(expiration_time):
        username = data.get("sub")
        if username:
            user = models.User.query.get(username)
            if user:
                # return user
                mfa_auth_time_string = data.get("mfa_auth_time")
                if mfa_auth_time_string:
                    mfa_auth_time = datetime.datetime.fromtimestamp(mfa_auth_time_string)
                    if mfa_auth_time >= dds_web.utils.current_time() - MFA_EXPIRES_IN:
                        return user

                if flask.request.path.endswith("/user/second_factor"):
                    return user

                # this should not be triggered more frequently than 15 mins
                #send_hotp_email(user)
                raise AuthenticationError(message="Two-factor authentication is required! "
                                                  "Please check your primary e-mail!")
        return None
    raise AuthenticationError(message="Expired token")
...
```

**Comment by worukan** [comment]
Also, in order to trigger MFA, I would have it as 
```
@basic_auth.verify_password
def verify_password(username, password):
    """Verify that user exists and that password is correct."""
    user = models.User.query.get(username)
    if user and user.verify_password(input_password=password):

        # this should not be triggered more frequently than 15 mins
        send_hotp_email(user)

        return user
    return None
```

File: requirements.txt
Comment (alneberg) [reply requirements.txt:37]: It's used in the `conftest.py`. But I have quite likely added it by mistake, I forgot we have a `requirements-test.txt` where this is already added. I'll remove it again.

Comment (alneberg) [reply requirements.txt]: Removed now.

**Comment by alneberg** [comment]
@worukan I see that your solution works as well. The benefit I can spot is that it would very easily allow for a separation between the token expiration and the mfa validity time.  I'ts not really hard from my implementation either though (one would just check the age of the token == time of mfa). 

The downside might be that it's slightly harder to add `TOTP` on top of yours since you directly trigger the hotp-mail upon requesting the token. With my solution a user would be able to send the `TOTP` request parameter directly to the token endpoint without first requesting the hotp token to be sent (since this is manually triggeded).

**Comment by worukan** [comment]
@alneberg I don't think you need to worry about using TOTP. I didn't outline the generic solution right now to save time and focus on the existing solution. I imagine it would depend on checking the user's mfa setup in the db and take the necessary action whether sending an email (for HOTP) or doing nothing (for TOTP). The server should know who uses HOTP or TOTP and act accordingly, and initiate when a user needs to authenticate with an MFA method.

File: dds_web/api/user.py
Comment (worukan) [reply dds_web/api/user.py:474]: For the previous code, update_token_with_mfa should be defined in dds_web/security/tokens.py as 

```
def update_token_with_mfa(token_claims):
    expires_in = datetime.datetime.fromtimestamp(token_claims.get("exp")) - dds_web.utils.current_time()
    return encrypted_jwt_token(username=token_claims.get("sub"),
                               sensitive_content=None,
                               expires_in=expires_in,
                               additional_claims={"mfa_auth_time": dds_web.utils.current_time().timestamp()})

```

File: dds_web/api/user.py
Comment (worukan) [reply dds_web/api/user.py:466]: For clarity, I meant that changes to encrypted token can be reverted and only the suggested line in the previous comment can be added.

File: dds_web/templates/mail/authenticate.html
Comment (worukan) [inline dds_web/templates/mail/authenticate.html:19]: Please, copy and paste the code below where you attempt to authenticate

File: dds_web/templates/mail/authenticate.txt
Comment (worukan) [inline dds_web/templates/mail/authenticate.txt:2]: Please, copy and paste the code below where you attempt to authenticate

File: dds_web/templates/mail/authenticate.html
Comment (worukan) [inline dds_web/templates/mail/authenticate.html:22]: Which token is this in here and authenticate.txt as well? Or can it be renamed to one_time_value or similar?

File: dds_web/templates/user/confirm2fa.html
Comment (worukan) [inline dds_web/templates/user/confirm2fa.html:5]: Not a significant issue, but would it be better to write something that a user can understand right away? 
Perhaps, DDS (either in acronym or full name) Multi-factor Authentication
I find it useful to have a clear title name for a browser tab which makes it easier to find it among many.

File: dds_web/templates/user/confirm2fa.html
Comment (worukan) [inline dds_web/templates/user/confirm2fa.html:21]: This sentence needs a rephrase and we should change 1 hour to 15 minutes both in text and code logic.

File: dds_web/templates/user/confirm2fa.html
Comment (alneberg) [reply dds_web/templates/user/confirm2fa.html:5]: Yes, I keep too many tabs to ever be able to read the title...

Comment (alneberg) [reply dds_web/templates/user/confirm2fa.html]: Fixed!

File: dds_web/templates/user/confirm2fa.html
Comment (alneberg) [reply dds_web/templates/user/confirm2fa.html:21]: That's a very strange sentence I agree. Ok, let's go with 15 min then.

Comment (alneberg) [reply dds_web/templates/user/confirm2fa.html]: Should be adjusted now.

File: dds_web/templates/mail/authenticate.html
Comment (alneberg) [reply dds_web/templates/mail/authenticate.html]: Changed it to `one_time_value`

File: dds_web/templates/user/confirm2fa.html
Comment (alneberg) [reply dds_web/templates/user/confirm2fa.html]: Done

File: dds_web/templates/mail/authenticate.html
Comment (worukan) [inline dds_web/templates/mail/authenticate.html:15]: I just noticed that _requested_ is in here as well. Let's change it to **issued**. It is in authenticate.txt as well.

File: dds_web/api/user.py
Comment (alneberg) [reply dds_web/api/user.py:473]: Not sure I understand the row `expires_in = datetime.datetime.fromtimestamp(token_claims.get("exp")) - dds_web.utils.current_time()`.  What change does that make to the expiration time of the token?

Comment (alneberg) [reply dds_web/api/user.py:473]: Nevermind I think I understand, it's because you add that time when you create the token in the other method.

**Comment by alneberg** [comment]
@worukan do you also have an implementation for the `send_hotp_email` or should I just put the two pieces together that I already have? Also, would you like me to implement a way to prevent the email to be sent more often than once every 15 min?

**Comment by worukan** [comment]
@alneberg 
I think we could place the following in security/auth.py
```
def send_hotp_email(user):
    hotp_value = user.generate_HOTP_token()
    msg = dds_web.utils.create_one_time_password_email(user, hotp_value)
    mail.send(msg)
```

I'm writing a small piece for the 15 mins frequency inside generate_HOTP_token. I will comment that in line in code soon.

File: dds_web/database/models.py
Comment (worukan) [inline dds_web/database/models.py:290]: could we change this to `hotp_issue_time` ?

File: dds_web/api/schemas/token_schemas.py
Comment (worukan) [inline dds_web/api/schemas/token_schemas.py:52]: Does data need to be returned from validate_mfa? What is the use of it? Currently it seems to be not used.

File: dds_web/database/models.py
Comment (worukan) [inline dds_web/database/models.py:413]: I think that lines 412 and 413 are not relevant and can be removed. Generate_HOTP_token already increments the counter before generating. Counter increment on line 412 doesn't happen on invalid one-time auth code and happens on successful verification. So on successful verification the counter is incremented twice.

File: dds_web/database/models.py
Comment (worukan) [inline dds_web/database/models.py:379]: It would be good to go over all code changes and replace occurrences of _password_ with **authentication code**. Also it is sometimes written with _one time_ and sometimes with **one-time**.

File: dds_web/api/schemas/token_schemas.py
Comment (worukan) [inline dds_web/api/schemas/token_schemas.py:48]: I have a suggestion for line 48. What do you think? Or could you make it even better?
The idea is if the entered otp value has expired or a wrong value is input and it has been more than 15 mins since otp has been issued, we will check and send a new email. As with the error, I didn't want to handle this in the model, so caught it only to raise again after sending email, if at all.

```
try:
    user.verify_HOTP(value.encode())
except AuthenticationError as error:
    dds_web.dds_web.security.auth.send_hotp_email(user)
    raise error
```

File: dds_web/database/models.py
Comment (worukan) [inline dds_web/database/models.py:398]: older than 15 minutes

File: dds_web/database/models.py
Comment (worukan) [inline dds_web/database/models.py:399]: Remember to update comment, if the increment will be removed from here

File: dds_web/web/user.py
Comment (worukan) [inline dds_web/web/user.py:160]: lines 154-160 look a little bit strange. It seems like you could rewrite it even as 

```
form.validate_on_submit()
redirect_to_login = flask.redirect(flask.url_for("auth_blueprint.login"))
        redirect_to_login.set_cookie("2fa_initiated_token", "", expires=0)
        return redirect_to_login
```

because no matter the outcome of validation you redirect to the same endpoint and might as well unset the cookie in any case?
And it seems like validate only returns a boolean, not that it raises an exception. Does it help us in any way?

File: dds_web/web/user.py
Comment (worukan) [inline dds_web/web/user.py:261]: Instead of lines 258-261, use 

```
if dds_web.dds_web.security.auth.send_hotp_email(user):
    flask.flash("One-Time Code has been sent to your primary email.")
```

In order for this to work, update in auth.py

```
def send_hotp_email(user):
    if not user.hotp_issue_time or (
        user.hotp_issue_time
        and (dds_web.utils.current_time() - user.hotp_issue_time > datetime.timedelta(minutes=15))
    ):
        hotp_value = user.generate_HOTP_token()
        msg = dds_web.utils.create_one_time_password_email(user, hotp_value)
        mail.send(msg)
        return True
    return False
```

**Review by inaod568 — COMMENTED** [review]
@worukan 
Apart from my added comments, I also find it strange to have the 15 min limit for the mfa when we have the limiter at 10 requests per hour for the basic authentication.

File: dds_web/api/schemas/token_schemas.py
Comment (worukan) [reply dds_web/api/schemas/token_schemas.py:48]: If you don't want this, it means that when a user's hotp verification fails and the hotp time expires, they have to do basic auth themselves again to be able to get a new hotp. This may not be obvious.

File: dds_web/web/user.py
Comment (worukan) [reply dds_web/web/user.py:191]: I think we are not interested in the underlying error messages such as invalid token or expired token. These do not make sense from a browser user's perspective. If there is something wrong with the token or it has expired, they have to supply the valid credentials and they will be redirected to the login page to do that.

File: dds_web/database/models.py
Comment (alneberg) [reply dds_web/database/models.py]: That's done by your PR already.

File: dds_web/api/schemas/token_schemas.py
Comment (inaod568) [reply dds_web/api/schemas/token_schemas.py:48]: I thought that's how it should work, at least for the CLI? But if not, could you write a list on how it should work for the web and cli? When someone needs to do basic auth and mfa.

File: dds_web/web/user.py
Comment (inaod568) [reply dds_web/web/user.py:191]: Sure, but for example if I were to log into a system with username and password, and then for some reason I wait too long to input the hotp, then I just get a message saying "You have to first supply valid credentials" - I would personally find this strange and think "but I did that, something must be wrong with this system". So I would like us to have a different message, maybe a concatenation of the underlying message and this message.

File: dds_web/api/schemas/token_schemas.py
Comment (alneberg) [reply dds_web/api/schemas/token_schemas.py:52]: Then I can remove the "mfa_verified" and "mfa_method" as well right? I'll go for that at least

File: dds_web/database/models.py
Comment (alneberg) [reply dds_web/database/models.py:410]: Without this, one could use the same HOTP token twice, as long as a new one has not been requested. While it might not be that easy to actually do it. It's very easy to use this to prevent it completely.

File: dds_web/database/models.py
Comment (alneberg) [reply dds_web/database/models.py]: I changed this occurrence. Searching for "time password" only found this one. Searching for "password" obviously finds more, so harder to check.

File: dds_web/web/user.py
Comment (alneberg) [reply dds_web/web/user.py]: I opted for Inas solution.

File: dds_web/web/user.py
Comment (alneberg) [reply dds_web/web/user.py:256]: Yes I think this is a good addition!

File: dds_web/database/models.py
Comment (inaod568) [reply dds_web/database/models.py:410]: Yes. Keep this.

File: dds_web/api/schemas/token_schemas.py
Comment (worukan) [reply dds_web/api/schemas/token_schemas.py:48]: OK, so I had a look at how the new cli behaviour is and it can prompt the user with "Second factor authentication failed, would you like to try again?". So I have a new suggestion, instead of making a change here, let's take `send_hotp_email(user)` from line 105 to line 101 in auth.py. This way a new value will be generated if needed, the first try to verify hotp will fail (it will fail anyway because it has been more than 15 mins already), but at least they get the email as hotp verify fails and the cli will prompt to try again. How about this?

File: dds_web/web/user.py
Comment (worukan) [reply dds_web/web/user.py:186]: perhaps "Supply valid credentials again due to timeout" or something like that

Comment (inaod568) [reply dds_web/web/user.py:186]: Yes something like that would improve a lot!

File: dds_web/api/schemas/token_schemas.py
Comment (inaod568) [reply dds_web/api/schemas/token_schemas.py:48]: Sounds good!

File: tests/test_basic_api.py
Comment (inaod568) [inline tests/test_basic_api.py:179]: Remove? Or adding more later?

**Review by inaod568 — APPROVED** [review]
🎉

**Comment by inaod568** [comment]
I'm approving after review and merging this now since @worukan had looked through and was ready to approve apart from two test related files.

**Pull request merged**
